api/src/main/java/org/apache/iceberg/expressions/StrictMetricsEvaluator.java

```diff
@@ -144,6 +146,51 @@ public Boolean or(Boolean leftResult, Boolean rightResult) {
       return ROWS_MIGHT_NOT_MATCH;
     }
 
+    @Override
+    @SuppressWarnings("checkstyle:CyclomaticComplexity")
+    public <T> Boolean isNaN(BoundReference<T> ref) {
+      int id = ref.fieldId();
+
+      if (nanCounts != null && nanCounts.containsKey(id) &&
+          valueCounts != null && valueCounts.containsKey(id)) {
+        if (nanCounts.get(id).equals(valueCounts.get(id))) {
+          return ROWS_MUST_MATCH;
+        }
+
+        return ROWS_MIGHT_NOT_MATCH;
+      }
+
+      // for v1 table, when NaN could still be upper/lower bound,
+      // if upper == lower == NaN and null count == 0, the column will only contain NaN
```

Comment:
Same here as above. If there are non-NaN values then this should return `ROWS_MIGHT_NOT_MATCH`.

Reply:
👍  will remove this block